test/test_ssl.cpp

```diff
@@ -611,13 +612,125 @@ void test_malicious_peer()
 	}
 }
 
+
+void test_ssl_magnet()
+{
+	error_code ec;
+	remove_all("tmp4_ssl", ec);
+	remove_all("tmp5_ssl", ec);
+
+	// set up session
+	int port = 1024 + rand() % 50000;
+	settings_pack sett = settings();
+	sett.set_int(settings_pack::max_retry_port_bind, 100);
+
+	char listen_iface[100];
+	std::snprintf(listen_iface, sizeof(listen_iface), "0.0.0.0:%ds", port);
+	sett.set_str(settings_pack::listen_interfaces, listen_iface);
+	sett.set_bool(settings_pack::enable_dht, false);
+	sett.set_bool(settings_pack::enable_lsd, false);
+	sett.set_bool(settings_pack::enable_upnp, false);
+	sett.set_bool(settings_pack::enable_natpmp, false);
+
+	lt::session ses2(session_params{sett, {}});
+	ses2.add_extension(&lt::create_ut_metadata_plugin);
+
+	port += 20;
+	std::snprintf(listen_iface, sizeof(listen_iface), "0.0.0.0:%ds", port);
+	sett.set_str(settings_pack::listen_interfaces, listen_iface);
+
+	lt::session ses1(session_params{sett, {}});
+	ses1.add_extension(&lt::create_ut_metadata_plugin);
+
+	wait_for_listen(ses1, "ses1");
+	wait_for_listen(ses2, "ses2");
+
+	// create torrent
+	create_directory("tmp4_ssl", ec);
+	std::ofstream file("tmp4_ssl/temporary");
+	std::shared_ptr<torrent_info> t = ::create_torrent(&file, "temporary"
+		, 16 * 1024, 13, false, {}, combine_path("..", combine_path("ssl", "root_ca_cert.pem")));
+	file.close();
+
+	TEST_CHECK(!t->ssl_cert().empty());
+
+	add_torrent_params addp;
+	addp.save_path = "tmp4_ssl";
+	addp.flags &= ~torrent_flags::paused;
+	addp.flags &= ~torrent_flags::auto_managed;
+	addp.ti = t;
+
+	torrent_handle tor1 = ses1.add_torrent(addp, ec);
+
+	tor1.set_ssl_certificate(
+		combine_path("..", combine_path("ssl", "peer_certificate.pem"))
+		, combine_path("..", combine_path("ssl", "peer_private_key.pem"))
+		, combine_path("..", combine_path("ssl", "dhparams.pem"))
+		, "test");
+
+	alert const* a = wait_for_alert(ses1
+		, torrent_finished_alert::alert_type, "ses1");
+	TEST_CHECK(a);
+	if (a)
+	{
+		TEST_EQUAL(a->type(), torrent_finished_alert::alert_type);
+	}
+
+	// create a new session with add torrent params using the root_certificate
+	create_directory("tmp4_ssl", ec);
+	add_torrent_params addp2;
+	addp2.save_path = "tmp5_ssl";
+	addp2.info_hashes = t->info_hashes();
+	addp2.flags &= ~torrent_flags::paused;
+	addp2.flags &= ~torrent_flags::auto_managed;
+
+	std::ifstream rootf(combine_path("..", combine_path("ssl", "root_ca_cert.pem")));
+	std::stringstream buffer;
+	buffer << rootf.rdbuf();
+	addp2.root_certificate = buffer.str();
+
+	torrent_handle tor2 = ses2.add_torrent(addp2, ec);
+
+	tor2.set_ssl_certificate(
+		combine_path("..", combine_path("ssl", "peer_certificate.pem"))
+		, combine_path("..", combine_path("ssl", "peer_private_key.pem"))
+		, combine_path("..", combine_path("ssl", "dhparams.pem"))
+		, "test");
+
+	ses1.listen_port();
+	ses2.listen_port();
```

Comment:
these calls appear meaningless. they return the listen port, right? and the return value is discarded

Reply:
Was copied from the [original test case](https://github.com/arvidn/libtorrent/blob/09754d7b627378cb1a215a2d964b4876ef20c35c/test/test_ssl.cpp#L236), I assume it's solving for some race condition; so I just cargo-culted it.

Reply:
I will add a comment